### ISSUE TITLE
False positive "Cannot move" on type that implements Copy

### ISSUE THREAD

**Issue by kluen** [intellij-rust/intellij-rust#3987]
## Environment

* **Intellij-Rust plugin version:**
v0.2.98.2125-191
* **Rust toolchain version:**
rustc 1.35.0 (3c235d560 2019-05-20)
* **IDE name and version:**
IntelliJ IDEA 2019.1.1 (Community Edition)
Build #IC-191.6707.61, built on April 16, 2019
JRE: 1.8.0_202-release-1483-b44 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
* **Operating system:**
macOS 10.14

## Problem description
Intellij shows a false positive "Cannot Move" error when trying to copy a `Point` value from the `nalgebra` library. This type implements `Copy` and therefore the given code compiles fine with rustc.

<img width="289" alt="Bildschirmfoto 2019-06-07 um 15 26 14" src="https://user-images.githubusercontent.com/13871367/59107312-a4594b00-8938-11e9-9e55-461856797f7e.png">

Unfortunately I was unable to narrow the case down to a minimal example without the library, but I included the necessary files to reproduce the error.

## Steps to reproduce
`Cargo.toml`:
```toml
[package]
name = "example"
version = "0.1.0"
edition = "2018"

[dependencies]
nalgebra = "0.18.0"
```

`main.rs`:
```rust
use nalgebra::base::dimension::U2;
use nalgebra::Point;

fn main() {}

fn test(x: &Point<f64, U2>) {
    let _y = *x;
}
```

**Comment by ortem** [comment]
```Rust
impl<N: Scalar, D: DimName> Copy for Point<N, D> where ... {}
impl<T: Copy + PartialEq + Debug + Any> Scalar for T {} // ==> impl Scalar for f64
```
Looks like our type inference cannot find `Copy` impl for `f64`. Example:

```Rust
#[derive(Clone)]
struct MyStruct<T> { t: T }

trait MyTrait {}

impl<T: MyTrait + Copy> Copy for MyStruct<T> {}
impl<T: Copy + PartialEq> MyTrait for T {}

fn test(x: MyStruct<f64>) {
    let _y = x;
    let _z = x; // false positive "Use of moved value"
}
```

**Comment by kluen** [comment]
Interestingly, with `0.2.102.2130-192`, @ortem's code above now is fine, whereas my initial example still shows an error. Seems to be a slightly different issue.

**Comment by foresterre** [comment]
An even more minimal example could be:

```rust
fn squared(a: i32) -> i32 {
    a * a
}
```

![image](https://user-images.githubusercontent.com/5955761/72689731-6ea25480-3b0c-11ea-834d-078f967791d4.png)

This is on `0.2.113.2155-192`

**Comment by kluen** [comment]
> This is on `0.2.113.2155-192`
I cannot confirm this on my machine (same version). Might be a different issue anyway?

**Comment by foresterre** [comment]
Ah, that's unfortunate; I felt it was the same issue based on the same error message, and similarly a Copy type. Apologies!

**Comment by Karrq** [comment]
I've been seeing this for a while now and it's getting increasingly annoying...
Is it being investigated?

**Comment by thechubbypanda** [comment]
This issue still exists prevalently, is it being looked into at all?
```
CLion 2021.1.2
Build #CL-211.7442.42, built on June 1, 2021
Linux 5.12.9-arch1-1
Non-Bundled Plugins: org.rust.lang (0.4.148.3911-211)
```
with
`rustc 1.52.1 (9bc8c42bb 2021-05-09)`

**Comment by David-OConnor** [comment]
Still happens.

**Comment by Maksim20023** [comment]
Hi @David-OConnor. Thanks for reaching us.
Please create a [new issue](https://github.com/intellij-rust/intellij-rust/issues/new) with a detailed description and code sample (or project) with described behavior. We'll look into the problem that you have encountered.

**Comment by canselcik** [comment]
So this is not acknowledged even though there are issues created all over the place?

This is most certainly happening still and very very annoying. Enough that I'd switch to something else if there were alternatives with no major shortcomings but there aren't.

I'm on 0.4.164.4409-212 which seems to be the latest.

**Comment by kluen** [comment]
Hi @canselcik, I'm the original poster.

I haven't used IntelliJ for Rust in a while, which is why I never checked if the bug is fixed.

I just checked with the following setup and the exact same files that I listed originally.

```
IntelliJ IDEA 2023.3 (Community Edition)
Build #IC-233.11799.241, built on December 2, 2023
Non-Bundled Plugins:
  org.rust.lang (233.15445)
```

Rust toolchain version is `1.75.0`.

The bug seems to be gone.

**Comment by canselcik** [comment]
Yeah, looks like CLion extension is no longer maintained however unless you
upgrade to the latest version of CLion, you will think that you are on the
latest version of the Rust plugin.

Once you upgrade to the latest CLion, you'll get the latest Rust plugin
update, telling you it is deprecated and that you should use RustRover.

Switching to RustRover fixes the issue.
